rfcs/20200103-grouped-convolution.md

```diff
@@ -0,0 +1,62 @@
+
+Grouped convolutions are now supported in TensorFlow core via [CUDNN 7](https://developer.nvidia.com/cudnn) on GPUs (https://github.com/tensorflow/tensorflow/pull/25818) and inside XLA compiled functions on CPU and GPUs, but there is no support via the Keras layers API yet. Adding support for grouped convolutions should make it easier for users to implement above mentioned models with TensorFlow and Keras in a fast way.
+
+## Design Proposal
+
+This proposal aims to add a `groups` argument to `keras.layers.Conv{1,2,3}D` which defines the number of channel groups similar to how other frameworks implement it in their high-level APIs.
+
+Since TensorFlow has native support for group convolutions, there are no changes required to the `keras.layers.Conv{1,2,3}D.call` methods, thus this proposal has no impact on the performance of standard convolutions in the Keras API and only requires small changes to `__init__` and `build` methods.
+
+```diff
+--- a/tensorflow/python/keras/layers/convolutional.py
++++ b/tensorflow/python/keras/layers/convolutional.py
+@@ -77,6 +77,16 @@ class Conv(Layer):
+       the dilation rate to use for dilated convolution.
+       Currently, specifying any `dilation_rate` value != 1 is
+       incompatible with specifying any `strides` value != 1.
++    groups: Integer, the number of channel groups controlling the connections
++      between inputs and outputs. Input channels and `filters` must both be
++      divisible by `groups`. For example,
++        - At `groups=1`, all inputs are convolved to all outputs.
++        - At `groups=2`, the operation becomes equivalent to having two
++          convolutional layers side by side, each seeing half the input
++          channels, and producing half the output channels, and both
++          subsequently concatenated.
++        - At `groups=input_channels`, each input channel is convolved with its
++          own set of filters, of size `input_channels / filters`
+     activation: Activation function to use.
+       If you don't specify anything, no activation is applied.
+     use_bias: Boolean, whether the layer uses a bias.
+@@ -106,6 +116,7 @@ class Conv(Layer):
+                padding='valid',
+                data_format=None,
+                dilation_rate=1,
++               groups=1,
+                activation=None,
+                use_bias=True,
+                kernel_initializer='glorot_uniform',
+```
+
+## Questions and Discussion Topics
+
+### Is it justified to become part of the core API?
+
+I think the amount of issues asking for this feature justifies an addition to the core API. The addition is also easy to maintain since it does not require large changes to the code base, but your mileage might vary. An alternative would be to add support for it to [TensorFlow Addons](https://github.com/tensorflow/addons), but this would introduce a substantial amount of code duplication between the repositories.
```

Comment:
What if in tensorflow addons, we subclass tf.keras.layers.Conv2D? Would that reduce the amount of code duplication?

Reply:
Thinking about this again, I am actually more concerned about the additional maintenance burden.
But, I also don't think it would reduce much duplication either, since one would need to implement a common base class used by 1D, 2D and 3D convolutions similar to the [private `Conv` layer](https://github.com/tensorflow/tensorflow/blob/c042ddcbb1e38569912f87f06ddcfb4e16b97d6a/tensorflow/python/keras/layers/convolutional.py#L48) which overwrites the [`build` method](https://github.com/tensorflow/tensorflow/blob/c042ddcbb1e38569912f87f06ddcfb4e16b97d6a/tensorflow/python/keras/layers/convolutional.py#L149-L189) which unfortunately relies on quite a few private methods.
The option with the least amount of code duplication would be to overwrite `Conv.add_weights` and changed the `kernel_shape` if `name=='kernel'`, though this seems quite brittle and hacky as well.

In general I think the main duplication either way would be all the docstrings and correct forwarding of arguments.

Reply:
@gabrieldemarmiesse Thinking about this again, it might be possible to implement this in TensorFlow Addons without relying on private implementation details and only calling `super()` methods, but I am not sure if it will fail due to some shape validation that needs to be changed as well. I would need to look closer into the implementation of it.

Reply:
Ok fair enough. As a side note, though, less maintenance isn't really an argument when modifying the public API (even if it can be a concern for other types of decisions). This is because we are modifying something that is visible to the users. And the importance of a good user experience is much, much, much higher than the maintenance cost. 

I mean, all those "compute_output_shape" that we have in the codebase are only for users to avoid specifying the input shape, like pytorch does, at each layer creating step, and save them some typing, and only in eager mode. That's a LOT of code for a nicer UX, but it's worth it. So, the philosophy of keras is UX >> maintainance costs.

I would suggest removing the duplication argument from the proposal as it won't influence the decision process. 

Afterwards, tensorflow addons was made exactly for this use case: a user proposes a new layer , a lot of people want to use it and we want to make the implementation available to everyone. Later on, we can decide to include it to core keras if we think it'll be used for years to come.

To bypass tensorflow addons, we would need a much stronger argument, because the layer your propose fits exactly in the processes described in tf.addons.

Reply:
> Ok fair enough. As a side note, though, less maintenance isn't really an argument when modifying the public API

Makes sense :+1:

> So, the philosophy of keras is UX >> maintainance costs

That's why Keras is such a joy to use 💯 

> I would suggest removing the duplication argument from the proposal as it won't influence the decision process.

Will do.

> Afterwards, tensorflow addons was made exactly for this use case: a user proposes a new layer , a lot of people want to use it and we want to make the implementation available to everyone. Later on, we can decide to include it to core keras if we think it'll be used for years to come.

That's a good point. The reason I propose to bypass TensorFlow Addons, is that it is an addition to an existing layer that has been requested for quite some time and is already supported by `tf.nn` and CUDNN. For me the main point to discuss is wether grouped convolutions are useful enough for many people to justify an addition to the core API.